### ISSUE TITLE
Performance loss and why?

### ISSUE THREAD

**Issue by DevelopDaily** [Dobiasd/frugally-deep#174]
The documentation has a section called "Performance" that compares the Keras + TF with Frugally-deep. Could you please interpret the results better?

I mean, C++ is supposed to be much faster than Python, but the Keras with Python beats Frugally-deep with C++ in most cases. Why? Is it caused by Frugally-deep not using GPU and using only one CPU? Well, GPU is everywhere in today's machines.  Why not to use it? Multi-core CPUs are prevalent too. Why not to use them?

I know this is not merely a technical question. Think about it from marketing's point of view. If your boss asks you why you should use Frugally-deep to suffer from performance loss in some cases, what can you say? What are the benefits of using Frugally-deep that are so clear and unique that you are willing to tolerate some performance loss?

**Comment by Dobiasd** [comment]
Hi, and thanks for your very valid questions.

The fact that you have them, indicates, that there might be room for improvement in the `README.md`, i.e., providing answers early on.

> The documentation has a section called "Performance" that compares the Keras + TF with Frugally-deep. Could you please interpret the results better?

The [performance section](https://github.com/Dobiasd/frugally-deep#performance) shows the average runtime of one single prediction on popular model types, using only one CPU and no GPU. Interpretation: Except for NASNet, there is no significant performance difference with this setup.

> I mean, C++ is supposed to be much faster than Python, but the Keras with Python beats Frugally-deep with C++ in most cases. Why? Is it caused by Frugally-deep not using GPU and using only one CPU?

That Python part of Keras and TensorFlow, that you interact with, is only the API. All the computational heavy stuff, happening under the hood, is mostly written in C++, as you can see in the [repo](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/core/kernels/conv_ops.cc). This is a common pattern for Python libraries. Numpy is another prominent example. All the heavy number crunching is written in C, Fortran or C++, and the Python API provides a convenient wrapper around those functions.

> Well, GPU is everywhere in today's machines. Why not to use it? Multi-core CPUs are prevalent too. Why not to use them?

It depends on the use-case. First, not all machines have GPUs, also in the cloud. Second, and maybe more importantly, in some use cases, it's not about latency (prediction time for a single input) but about throughput, i.e., how many predictions can you do per second overall. And if you have 20 CPU cores, you can run 20 predictions with frugally-deep in parallel on that machine. Since the model is thread-safe, you only need to load it once into the RAM.
Also, considering costs (buying hardware and using it, power consumption, etc.), there might be some cases where it's even cheaper to use CPUs instead of GPUs. Or some company might already have a Kubernetes cluster available with some free CPU resources, so it's more cost-efficient to just utilize them. GPU prediction brings the biggest advantage when you use batch prediction, but some in some use cases, it might be simpler to implement and to maintain, to not accumulate batches, but just run a single prediction when needed.

And, coming back to latency once more, not all models scale perfectly fine to multiple cores. Even one single big matrix multiplication (which is basically what happens in a convolutional layer), does not become 10 times faster just because you use 10 cores in parallel. You will fill them all with work, but maybe your prediction only becomes 5 times faster. Overall, you would have more throughput if you just run 10 predictions in parallel, one on each core.

> If your boss asks you why you should use Frugally-deep to suffer from performance loss in some cases, what can you say?

Independent of the answers above: If you have a reason why you have chosen frugally-deep for your project, and you need to defend your decision to your boss, just explain to him the reasons you used to convince yourself. If you don't have good reasons to use frugally-deep, you should not use it. Nobody forces you. :slightly_smiling_face:

> What are the benefits of using Frugally-deep that are so clear and unique that you are willing to tolerate some performance loss?

As described above, performance is use-case dependent.

Also, I can re-phrase your question as an analogy:

"Why should I use a bicycle instead of an SUV?"

Unique things, that set frugally-deep apart from TensorFlow:

- It works on 32 bit operating systems. This might be important for legacy systems or embedded development.

- It's header-only. This means it's much simpler to integrate into your software. No need to compile external `dll`/`lib`/`so` files. Also the size of the application you ship will be much smaller.

So the decision really depends on the use-case. Frugally-deep is not only used by private people or by scientists, but also by some larger company-projects. But it's intent is not to replace TensorFlow everywhere. It's just an option you can freely choose to consider.

And finally, regarding features, keep in mind, that TensorFlow is a large project, developed by many very good people, paid by Google. Frugally-deep, on the other hand, is mostly written and maintained by me, in my spare time, plus some external contributions. If you see a use-case for it, I'm very happy if you find it helpful. If not, I'm also totally OK with it.

**Comment by DevelopDaily** [comment]
Your work is commendable. The answers are clear and would be beneficial to other newcomers like me. Thank you.

I am very glad you have made the design decisions to offer the 32-bit option and header-only implementation. A lot of innovative applications still need the 32-bit.

**Comment by Dobiasd** [comment]
Thanks. I [adjusted the `REAMDE.md`](https://github.com/Dobiasd/frugally-deep/commit/2a6360eb19b95aa2952f66f63bdc6241533a385c) to emphasize the performance-test setup more. Also, it now mentions the option of parallel predictions.

**Comment by DevelopDaily** [comment]
Great. The case of Frugally-deep is much more convincing now.

A little typo here "...utilizing as many CPUs are you like..."

Did you mean to say "...as you like..."?

**Comment by Dobiasd** [comment]
Right. Thanks a lot. :+1:
I just fixed it.
